File: java/high-scores/src/main/java/HighScores.java
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

class HighScores {

    private List<Integer> scores;

    public HighScores(List<Integer> highScores) {
        this.scores = highScores;
    }

    List<Integer> scores() {
        return this.scores;
    }

    Integer latest() {
        return this.scores.get(this.scores.size() - 1);
    }

    Integer personalBest() {
        return scores.stream().max(Comparator.naturalOrder()).orElse(0);
    }

    List<Integer> personalTopThree() {
        return scores.stream()
                .sorted(Comparator.reverseOrder())
                .limit(3)
                .collect(Collectors.toList());
    }

}

File: solutions/lua/high-scores/1/high-scores.lua
local HighScores = {}

function HighScores:scores()
  return self._scores
end

function HighScores:latest()
  return self._scores[#self._scores]
end

function HighScores:personal_best()
  sorted = {}
  for i=1, #self._scores do
    sorted[i] = self._scores[i]
  end

  table.sort(sorted)

  return sorted[#sorted]
end

function HighScores:personal_top_three()
  sorted = {}
  for i=1, #self._scores do
    sorted[i] = self._scores[i]
  end

  table.sort(sorted)

  return {sorted[#sorted], sorted[#sorted - 1], sorted[#sorted - 2]}
end

return function(scores)
  local high_scores = {}
  setmetatable(high_scores, { __index = HighScores })

  high_scores._scores = scores

  return high_scores
end
